File: demos/scala/tcp-server-client/src/test/scala/TCPServerTest.scala
import java.util.concurrent.{Executors, TimeUnit}

import org.scalatest.{Matchers, FlatSpec, BeforeAndAfter}

import scala.collection.mutable.ListBuffer

import com.github.pheymann.mockit.MockIt
import com.github.pheymann.mockit.core.{ConnectionType, ServerConfiguration, ShutdownLatch}
import com.github.pheymann.mockit.annotation.{MockItConfig, MockItConfigs}

@MockItConfigs
class TCPServerTest extends FlatSpec with BeforeAndAfter with Matchers {

    final val PORT  = 1234

    val serverIps   = new ListBuffer[(String, String, Int)]
    val pool        = Executors.newSingleThreadExecutor

    var shutdown: ShutdownLatch = null

    @MockItConfig(mockKey = "test", mockUnit = "Multiply2ServerMock")
    val testConfig = new ServerConfiguration(
        PORT,
        1,
        ConnectionType.tcp
    )
    @MockItConfig(mockKey = "test", mockUnit = "Multiply3ServerMock")
    val testConfig2 = new ServerConfiguration(
        PORT + 1,
        1,
        ConnectionType.tcp
    )

    before {
        val latch = new ShutdownLatch

        shutdown = new ShutdownLatch

        pool.submit(MockIt.mockNetwork("test", "", serverIps, shutdown,latch))
        latch.await()
    }

    after {
        serverIps.clear()

        shutdown.close
        pool.shutdown()
        pool.awaitTermination(1, TimeUnit.MINUTES)
    }

    "A MultiplyServer" should "multiply a number provided by a client" in {
        val x = 4
        val y = 5

        var factor  = 0

        for ((mock, ip, port) <- serverIps) {
            val client = new ClientAction(ip, port)

            if (mock.contains("Multiply2"))
                factor = 2
            else
                factor = 3

            (x * factor + y * factor) should be (client.action(x, y))
            (x * factor + y * factor) should be (client.action(x, y))
        }
    }

}

File: demos/java/tcp-server-client/src/test/java/TCPServerTest.java
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import com.github.pheymann.mockit.core.Configuration;
import com.github.pheymann.mockitjavaapi.core.ConfigurationFactory;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import static org.junit.Assert.*;

import com.github.pheymann.mockitjavaapi.General;
import com.github.pheymann.mockitjavaapi.JMockIt;
import com.github.pheymann.mockitjavaapi.core.JMockCallable;
import com.github.pheymann.mockit.core.ShutdownLatch;
import com.github.pheymann.mockit.annotation.MockItConfig;
import com.github.pheymann.mockit.annotation.MockItConfigs;

@MockItConfigs
public class TCPServerTest {

    private final int               PORT        = 1234;

    private final ExecutorService   pool        = Executors.newSingleThreadExecutor();

    private JMockCallable           mockJob     = null;
    private List<Object[]>          serverIps   = null;

    private ShutdownLatch           shutdown    = new ShutdownLatch();

    @MockItConfig(mockKey = "test", mockUnit = "Multiply2ServerMock")
    private final Configuration testConfig = ConfigurationFactory.createServer(
        PORT,
        1,
        General.JConnectionType.tcp
    );
    @MockItConfig(mockKey = "test", mockUnit = "Multiply3ServerMock")
    private final Configuration testConfig2 =ConfigurationFactory.createServer(
        PORT + 1,
        1,
        General.JConnectionType.tcp
    );

    @Before
    public void init() throws Exception {
        final ShutdownLatch latch = new ShutdownLatch();

        shutdown    = new ShutdownLatch();
        mockJob     = JMockIt.mockNetwork("test", "", shutdown, latch);

        pool.submit(mockJob);
        latch.await();

        serverIps   = mockJob.getServerIps();
    }

    @After
    public void delete() throws Exception {
        serverIps.clear();

        shutdown.close();
        pool.shutdown();
        pool.awaitTermination(1, TimeUnit.MINUTES);
    }

    @Test
    public void testTCPServer() throws Exception {
        final int x = 4;
        final int y = 5;

        int factor  = 0;

        for (Object[] serverIp : serverIps) {
            final ClientAction client = new ClientAction((Integer) serverIp[2], (String) serverIp[1]);

            if (((String) serverIp[0]).contains("Multiply2"))
                factor = 2;
            else
                factor = 3;

            assertEquals(x * factor + y * factor, client.action(x, y));
            assertEquals(x * factor + y * factor, client.action(x, y));
        }
    }

}
